### PR TITLE
fix(tasks): bound the OOM-crash GET tasks query — Phase 2 of task board crash-prevention epic

### PR DESCRIPTION
**Pull request by 2witstudios**
## Summary

Production Postgres OOM-crashed on 2026-07-18 (17:30–17:45 UTC). Root cause, confirmed from Postgres logs and code inspection: `apps/web/src/app/api/pages/[pageId]/tasks/route.ts`'s GET handler ran `db.query.taskItems.findMany` scoped only by `inArray(taskItems.pageId, childPageIds)` — no limit/offset — joining 5 relations (`assignee`, `assigneeAgent`, `user`, `page`, `assignees`). Sorting and search-filtering then happened in JS *after* the unbounded fetch. A backend running this query hit 3.5GB RSS on a 4GB machine and got OOM-killed, forcing full Postgres crash recovery. This is Phase 2 of the "Task Board Query & Reorder Safety" epic ([PageSpace page j44e35jwzlhr54fbmruk3k4i](https://app.pagespace.ai)).

- **`query-spec.ts`** (new): `parseTaskQuerySpec`, a pure parser that clamps `limit` (1–200, default 100) and `offset` (>= 0, default 0) via the existing `parseBoundedIntParam` helper, and falls back to `sortOrder: 'asc'` for unrecognized values. Status/assigneeId/search pass through unmodified. Also exports `escapeLikePattern`, escaping `%`/`_`/`\` so a search term is matched as a literal substring rather than an ILIKE pattern.
- **`route.ts`**: the GET handler now runs two bounded queries instead of one unbounded one:
  1. A lightweight join (`taskItems` ⋈ `pages`) resolves the ordered — by `pages.position` coalesced with `taskItems.position`, preserving the exact prior sort semantic — filtered (status/assigneeId via `where`, search via escaped `ilike` on `pages.title`), limited/offset set of task **ids**.
  2. The existing relational `findMany` (unchanged shape/columns/relations) hydrates only those bounded ids via `inArray(taskItems.id, ids)`, with an explicit `limit: ids.length` as defense-in-depth (see below).
- Response shape is unchanged — `TaskListView.tsx`/`TaskKanbanView.tsx` need no frontend changes.
- Does not touch the reorder route (separate phase of the epic).

## Changes made during review convergence

Automated review (Codex) plus a self-run 4-angle `/simplify` pass surfaced real issues, addressed here:

1. **LIKE metacharacter escaping (P2, fixed):** a search term containing a literal `%`, `_`, or `\` was passed unescaped into `ilike()`, so Postgres read it as pattern syntax instead of literal text. Added `escapeLikePattern`.
2. **Cross-PR lint hazard (self-found, fixed):** sibling Phase 8 (PR #2129, open in parallel) adds an ESLint rule requiring every `db.query.*.findMany()` to carry a literal `limit` key. This PR's phase-2 hydrate call was bounded only indirectly via `inArray`. Added `limit: boundedTaskIds.length` so it stays compliant and self-evidently bounded regardless of merge order.
3. **Frontend pagination cap (P1, addressed via reply, not code change):** Codex flagged that task lists with >100 children will now silently show only the first page with no load-more UI. Verified against the epic's own Phase 2 task spec on the PageSpace board, which explicitly states "no frontend changes required" — this is a deliberate, already-documented scope boundary for this 8-phase, backend-only epic, not an oversight. Logged as a separate follow-up task rather than expanding this PR's scope.
4. **Simplification pass:** deduped the identical empty-response object literal (built twice, once per early-return branch) into a local `emptyTasksResponse()` closure; flattened `parseTaskQuerySpec`'s triple conditional-spread into direct `x || undefined` assignments.
5. **Logged, not fixed here (real but out of scope):** `escapeLikePattern` duplicates 3 existing local implementations elsewhere in `apps/web` (`api/tasks/route.ts`, `api/search/route.ts`, `api/mentions/search/route.ts`). Centralizing means touching 3 files outside this phase's declared scope (`route.ts` + `query-spec.ts` only) — logged as a follow-up task.
6. **Non-deterministic pagination on tied positions (self-found via a final independent review pass, fixed):** `ORDER BY coalesce(pages.position, taskItems.position)` had no secondary sort key. Before this PR the route fetched every task unconditionally and sorted in JS, so ties were invisible; once phase 1 applies `LIMIT`/`OFFSET` in SQL, two tasks sharing a position (e.g. a read-then-write race in `POST`'s `nextPosition` computation, or a backfilled row that never got a distinct position) have no guaranteed stable order across repeated calls — paging could silently skip or duplicate rows. Added `taskItems.id` as a tiebreaker in the `ORDER BY`.

## Cross-PR verification

This epic had 4 sibling PRs in flight in parallel (#2126 Phase 7, #2127 Phase 1, #2129 Phase 8 — all merged to `master` during this PR's review; #2130 Phase 3 still open). Phase 8's merge produced a real conflict in `route.ts` (its ESLint rule annotated the exact unbounded `findMany` this PR replaces) — resolved by keeping this PR's bounded rewrite, which already satisfies the new lint rule via its explicit `limit` key; branch is merged with current `master` and `mergeable: MERGEABLE` / `mergeStateStatus: CLEAN`. `parseBoundedIntParam`'s import path (`@/lib/utils/query-params`) is preserved as a re-export shim by Phase 1's relocation, so no further changes were needed there.

## Test plan

- [x] `query-spec.test.ts`: 21 tests, 100% branch coverage of `parseTaskQuerySpec` and `escapeLikePattern` (defaults, limit/offset clamping in both directions, invalid sortOrder fallback, each filter param present/absent, LIKE metacharacter escaping).
- [x] Extended `route.test.ts` (42 tests) with regressions for: results capped at the query limit with hydration scoped to the bounded ids, search narrowing happening via SQL not JS, LIKE metacharacters escaped before reaching `ilike()`, the defense-in-depth `limit` key on the hydrate call, and the `taskItems.id` tiebreaker on the phase-1 `orderBy`.
- [x] `bun run typecheck` (root, full monorepo) — green (16/16 turbo tasks), re-verified after the master merge.
- [x] `bun run lint` (apps/web) — clean, no new warnings/errors.
- [x] `bun run test` from `apps/web` — 990 files / 14929 tests passed (3 pre-existing failing files unrelated to this change: a missing DB role affecting `admin-role-version.test.ts`/`activity-tools.test.ts`, and a timezone edge case in `grouping.test.ts` — none touch the tasks route).
- [x] No merge conflicts with `master` (merged and pushed; `gh pr view` reports `mergeable: MERGEABLE` / `mergeStateStatus: CLEAN`).
- [x] All GitHub CI checks green: Unit Tests, Lint & TypeScript Check, Security Test Suite, Dependency Audit, Secret Scanning, Static Security Analysis, CodeQL, CodeRabbit.
- [x] Independent final correctness/security review pass (separate from the `/simplify` pass) found one real issue (the tiebreaker above, now fixed) and confirmed no auth/scope-leak risk, no SQL injection risk, and no off-by-one in limit/offset clamping.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01R9HNEnE2gFPs1c7UrjJGcf

### CODE REVIEW

File: apps/web/src/app/api/pages/[pageId]/tasks/route.ts [hunk 1]
```diff
@@ -144,13 +142,43 @@ export async function GET(req: Request, { params }: { params: Promise<{ pageId:
...
+  const orderedIdRows = await db
+    .select({ id: taskItems.id })
+    .from(taskItems)
+    .innerJoin(pages, eq(pages.id, taskItems.pageId))
+    .where(and(
       inArray(taskItems.pageId, childPageIds),
       status ? eq(taskItems.status, status) : undefined,
       assigneeId ? eq(taskItems.assigneeId, assigneeId) : undefined,
-    ),
+      search ? ilike(pages.title, `%${search}%`) : undefined,
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/pages/[pageId]/tasks/route.ts]: Good catch — this was a real bug. Fixed in the latest commit: added `escapeLikePattern` (a pure helper in `query-spec.ts`) that escapes `%`, `_`, and the backslash escape character itself before the search term is interpolated into the `ilike(pages.title, ...)` pattern, so a title like "100% done" matches literally instead of `%` being read as a wildcard. Covered by 5 new unit tests on the escaping function plus a route-level regression test asserting the exact escaped pattern reaches `ilike()`.

File: apps/web/src/app/api/pages/[pageId]/tasks/route.ts [hunk 2]
```diff
@@ -144,13 +142,43 @@ export async function GET(req: Request, { params }: { params: Promise<{ pageId:
...
+    ))
+    .orderBy(sortOrder === 'desc' ? desc(positionExpr) : asc(positionExpr))
+    .limit(limit)
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/pages/[pageId]/tasks/route.ts:165]: Valid observation, but this is a deliberate, documented scope boundary rather than an oversight. This PR is Phase 2 of an 8-phase epic ("Task Board Query & Reorder Safety") tracked in our PageSpace board — the Phase 2 task spec explicitly states: *"Given the current no-params request shape, should return the same response fields/shape as today (no frontend changes required — `TaskListView.tsx` / `TaskKanbanView.tsx` fetch once and filter/sort client-side over the full — now bounded — result)."* All 8 phases are backend-only (bounded queries, deadlock-safe reorder primitive, DB-level statement/lock timeouts, an ESLint backstop for unbounded `findMany`) — none touch frontend pagination/list UI.

The tradeoff you flagged is real: a task list with more than `limit` (default 100, max 200) direct children will silently show only the first page today. That is an accepted cost of this phase given the alternative — an unbounded query that OOM-crashed production — and the practical risk is bounded (100 covers the overwhelming majority of real task lists; the incident itself was caused by an unbounded fetch, not normal-scale data).

Adding "load more"/pagination to `TaskListView`/`TaskKanbanView` is real, worthwhile follow-up work, but it is a separate UI concern from this epic and out of scope for this PR. I will flag it to the epic owner as a candidate for a follow-up task rather than expanding this PR's scope.
